options/options.go

```diff
@@ -1497,6 +1723,130 @@ func (interp *interpreter) scalarFieldValue(mc *internal.MessageContext, fldType
+		if opt.StringValue != nil {
+			return opt.GetStringValue(), nil
+		}
+		return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vexpecting bytes, got %s", mc, optionValueKind(opt))
+	case descriptorpb.FieldDescriptorProto_TYPE_STRING:
+		if opt.StringValue != nil {
+			return string(opt.GetStringValue()), nil
+		}
+		return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vexpecting string, got %s", mc, optionValueKind(opt))
+	case descriptorpb.FieldDescriptorProto_TYPE_INT32, descriptorpb.FieldDescriptorProto_TYPE_SINT32, descriptorpb.FieldDescriptorProto_TYPE_SFIXED32:
+		if opt.NegativeIntValue != nil {
+			i := opt.GetNegativeIntValue()
+			if i > math.MaxInt32 || i < math.MinInt32 {
+				return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vvalue %d is out of range for int32", mc, i)
+			}
+			return int32(i), nil
+		}
+		if opt.PositiveIntValue != nil {
+			ui := opt.GetPositiveIntValue()
+			if ui > math.MaxInt32 {
+				return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vvalue %d is out of range for int32", mc, ui)
+			}
+			return int32(ui), nil
+		}
+		return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vexpecting int32, got %s", mc, optionValueKind(opt))
+	case descriptorpb.FieldDescriptorProto_TYPE_UINT32, descriptorpb.FieldDescriptorProto_TYPE_FIXED32:
+		if opt.NegativeIntValue != nil {
+			i := opt.GetNegativeIntValue()
+			if i > math.MaxUint32 || i < 0 {
+				return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vvalue %d is out of range for uint32", mc, i)
+			}
+			return uint32(i), nil
+		}
+		if opt.PositiveIntValue != nil {
+			ui := opt.GetPositiveIntValue()
+			if ui > math.MaxUint32 {
+				return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vvalue %d is out of range for uint32", mc, ui)
+			}
+			return uint32(ui), nil
+		}
+		return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vexpecting uint32, got %s", mc, optionValueKind(opt))
+	case descriptorpb.FieldDescriptorProto_TYPE_INT64, descriptorpb.FieldDescriptorProto_TYPE_SINT64, descriptorpb.FieldDescriptorProto_TYPE_SFIXED64:
+		if opt.NegativeIntValue != nil {
+			return opt.GetNegativeIntValue(), nil
+		}
+		if opt.PositiveIntValue != nil {
+			ui := opt.GetPositiveIntValue()
+			if ui > math.MaxInt64 {
+				return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vvalue %d is out of range for int64", mc, ui)
+			}
+			return int64(ui), nil
+		}
+		return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vexpecting int64, got %s", mc, optionValueKind(opt))
+	case descriptorpb.FieldDescriptorProto_TYPE_UINT64, descriptorpb.FieldDescriptorProto_TYPE_FIXED64:
+		if opt.NegativeIntValue != nil {
+			i := opt.GetNegativeIntValue()
+			if i < 0 {
+				return nil, reporter.Errorf(interp.nodeInfo(node).Start(), "%vvalue %d is out of range for uint64", mc, i)
+			}
+			// should not be possible since i should always be negative...
+			return uint64(i), nil
+		}
```

Comment:
While we wouldn't expect to see it in a properly-formed file descriptor set, it is technically possible for `NegativeIntValue` to be non-nil and explicitly set to zero (or a positive value). So I think we still need both branches that are here as a just in case.